File Access fix

daschuer opened the pull request:
Due to refactoring the thread safe access of the Tracks File access object is no longer thread safe. 
https://github.com/mixxxdj/mixxx/blob/2.4/src/track/track.h#L82

I have removed this function and pass the thread safe TrackPointer instead. 
This has the big advantage, that we can use it directly with the sound source for extracting the cover and don't need to use the location of the track outside the track locking it up again from the Track cache later. 

In https://github.com/mixxxdj/mixxx/issues/11131 we have seen that the independent track location was empty in the concurrent case. 

I don't think that getFileAccess() was the cause or even a real Problem, because it is almost all the time read only, but hopefully the additional ASSERTS and the streamlined control flow helps to find the real root cause.

(I am still not able to produce the issue) 

@JoergAtGithub can you check this out with debug asserts fatal?

daschuer committed:
Don't pipe unused FileAccess object through relocateCachedTrack()

daschuer committed:
guessCoverInfoForTrack() takes a TrackPointer to avoid a track cash roundtrip

daschuer committed:
Introduce extractEmbeddedCover(pTrack)

daschuer committed:
Remove unsafe Track::getFileAccess() and introduce getFileAccessToken()

daschuer committed:
Allow to use a TrackPointer CoverInfo::loadImage

daschuer committed:
Assert that the pTrack matches with the CoverInfo object

JoergAtGithub commented:
I did a short test. It worked without assert. I noticed, that the spinny coverart is now loaded only very seldom (once in 20 track loads), otherwise everything feels as before.

daschuer commented:
That is good news. 
We know now that the issue is fixed with a lock and got worse without any lock. 

I have already tried to make a unit-test, but it was always good. Maybe I can make it fail now. 

I suggest to merge this PR, because it makes the issue more likely, so it will be easier to verify a potential fix of the real root cause.

m0dB commented:
@daschuer , I am not familiar enough with this code to have a useful opinion about this PR, and it is not entirely clear to me what the objective is, but I will can try to run it on macOS with thread sanitizer to see if flags up some issue. Would that help?

daschuer commented:
I could drill that down into Taglib near https://github.com/taglib/taglib/blob/master/taglib/toolkit/tfilestream.cpp#L57. But I am not able to explain it. Everything looks good. 

The symptoms are that the FileName std::wstring which is copied around many times, is loosing its value becomming null. 
LastError is Null as well. Maybe an outer class is reinitialized or such. 

It happens whenever the same file is accessed at the same time from different threads.

The poor mans solution would be to guard that region with a mutex. 

There is on the other hand no reason to access the same file from different threads at the same time. So we should in any case think about improving the coverart cache to sort this out. I have something in progress. 

But I am really curious to find the root cause to not have another symptom popping up elsewhere later.

Debugging on Linux and MacOS does not make any sense, because that code is windows only.

daschuer commented:
A first step would be to merge this which makes the issue more likely. Maybe a windows contributor can than take-over. 
I will continue in the meantime to improve the cover-art cache.

m0dB commented:
Sounds good! If there is anything I can do on macOS, let me know. In the meantime, I did try to make a thread sanitizer enabled build: Independent of this bug, we definitely should run with thread sanitizer. Unfortunately, I immediately ran into false positives with QMutex. I will take the discussion about that to zulip.

Swiftb0y commented on src/track/globaltrackcache.cpp:
```diff
@@ -341,15 +341,13 @@ void GlobalTrackCache::relocateTracks(
                 trackRef = TrackRef::fromFileInfo(
                         relocatedFileAccess.info(),
                         trackRef.getId());
-                fileAccess = std::move(relocatedFileAccess);
             }
```
why the removal? this line looks crucial to me.

daschuer replied:
`fileAccess` has been removed completely from this function.  The new file access is already owned by the track after  `plainPtr->relocate()`. So there is no use for this code line anymore.

Swiftb0y replied:
Ahh sorry, I was only looking at this line in the context of the commit which still contained the fileAccess variable above.

Swiftb0y commented on src/library/coverartutils.cpp:
```diff
@@ -196,46 +196,46 @@ CoverInfoRelative CoverInfoGuesser::guessCoverInfo(
-        const Track& track) {
-    const auto fileAccess = track.getFileAccess();
+        TrackPointer pTrack) {
+    const auto fileAccess = pTrack->getFileAccess();
```
how does this avoid a TrackCache roundtrip?

Swiftb0y commented on src/library/coverartutils.h:
```diff
@@ -30,6 +30,8 @@ class CoverArtUtils {
     // Extracts the first cover art image embedded within the file.
     static QImage extractEmbeddedCover(
             mixxx::FileAccess trackFileAccess);
+    static QImage extractEmbeddedCover(
+            TrackPointer pTrack);
```
whats the advantage of this overload over the `FileAccess` version?

daschuer replied:
If we have a track, we can us it directly and bypass the track lookup logic.

Swiftb0y replied:
What do we gain from that?

Swiftb0y commented on src/library/coverartutils.cpp:
```diff
@@ -56,6 +56,18 @@ QImage CoverArtUtils::extractEmbeddedCover(
+QImage CoverArtUtils::extractEmbeddedCover(
+        TrackPointer pTrack) {
+    QImage image;
+    // Both resetMissingTagMetadata = false/true have the same effect
+    constexpr auto resetMissingTagMetadata = false;
```
that sounds like a bug?

daschuer replied:
I did not investigate that. The comment is a literal copy form the previous version.

uklotzde replied:
If you don't read any other metadata by passing `nullptr` the option has no effect.

Swiftb0y commented on src/track/track.h:
```diff
@@ -78,14 +78,16 @@ class Track : public QObject {
-        // Copying QFileInfo is thread-safe due to implicit sharing,
+        // Copying mixxx::FileInfo based on QFileInfo is thread-safe due to implicit sharing,
         // i.e. no locking needed.
+        static_assert(sizeof(QSharedDataPointer<int>) == sizeof(mixxx::FileInfo));
```
while I appreciate the usage of the `static_assert`, I'm not sure the way its implemented is a good idea. It's very leaky. 
I think it would be better if there was a `constexpr static` member on `mixxx::FileInfo` that indicated whether it only consisted of implicitly shared types. Ideally Qt had some machinery for this, but I don't think it does.

daschuer replied:
I have put it here because it proves the comment just in the next line.  If one will add members to FileInfo it will fail here where the mutex lock is required.

Swiftb0y replied:
Right, but it will not fail if the implementation changes but it happens to have the same size. That's why I proposed to have the evaluation of the property in the fileaccess class, but leave the static_assert here. It's just shifting code around a bit in order to not break encapsulation.

daschuer replied:
Now we have a static isQFileInfo(). I was attempted to call it isThreadSafeCopyable() but that would be very confusing in a Info class describing file actions.

Swiftb0y replied:
>  I was attempted to call it isThreadSafeCopyable() but that would be very confusing in a Info class describing file actions.

Yeah, but its a naming issue. The entire point of the static is not to expose the internal QFileInfo imo. I think it sufficies to use a better name or doc comment.

Swiftb0y commented on src/track/track.h:
```diff
@@ -78,14 +78,16 @@ class Track : public QObject {
-        // Copying QFileInfo is thread-safe due to implicit sharing,
+    SecurityTokenPointer getFileAccessToken() const {
+        // Copying a QSharedPointer is thread-safe
         // i.e. no locking needed.
```
why was this unsafe previously? `FileAccess` consists of a `FileInfo` and a `QSharedPointer`, both are thread-safe to copy.

daschuer replied:
Yes, but the context switch may happen in between so that the you have the old FileInfo with a new access token.

Swiftb0y commented on src/library/coverart.cpp:
```diff
@@ -106,19 +106,20 @@ QDebug operator<<(QDebug dbg, const CoverInfoRelative& info) {
             return loadedImage;
         }
         loadedImage.location = trackLocation;
-        loadedImage.image = CoverArtUtils::extractEmbeddedCover(
-                mixxx::FileAccess(
-                        mixxx::FileInfo(trackLocation),
-                        pTrackLocationToken));
```
won't this cause problems on environments with stricter security such as MacOS since we don't have a token anymore?

daschuer replied:
We till have the token. In case we have a Track, the track owns the access token. 
In the other case a new access token is requested like before where pTrackLocationToken was null.

Swiftb0y replied:
Okay, that second behavior was not obvious. Makes sense.

daschuer commented:
It turns out that the real bug is a weakness of the "CreateFileW()" the first thread is opening the file which seems to lock it. 
The second thread tries to access the same file R/W which fails than it tries again in RO mode which also fails and gives up finally.

daschuer committed:
move the type check into the class definition

Swiftb0y commented on src/util/fileinfo.h:
```diff
@@ -255,6 +257,13 @@ class FileInfo final {
+        // Additional member variables will violate this assumption
+        return (sizeof(QFileInfo) == sizeof(mixxx::FileInfo) &&
+                std::is_same_v<decltype(m_fileInfo), QFileInfo>)
+    }
```
```suggestion
    // This only works because Qt's implicit sharing allows copies to be threadsafe and
    // this class only consists of a single QFileInfo member.
    // Adding additional member variables will likely violate this assumption.
    static constexpr bool canBeCopiedThreadSafe = sizeof(QFileInfo) == sizeof(mixxx::FileInfo) && std::is_same_v<decltype(m_fileInfo), QFileInfo>
    }
```

Swiftb0y commented on src/track/track.h:
```diff
@@ -81,7 +81,7 @@ class Track : public QObject {
         // Copying mixxx::FileInfo based on QFileInfo is thread-safe due to implicit sharing,
         // i.e. no locking needed.
-        static_assert(sizeof(QSharedDataPointer<int>) == sizeof(mixxx::FileInfo));
+        static_assert(mixxx::FileInfo::isQFileInfo());
```
```suggestion
        static_assert(mixxx::FileInfo::canBeCopiedThreadSafe);
```

daschuer replied:
As I have described above, I had this original but picked isQFileInfo because than it is obvious that it is not related to the file itself. It is intended to expose the QFileInfo nature here, because than the developer can read through the QT docs.
It is a difference to expose something to the developer by a name, compared to exposing implementation details to the compiler. The later is now fixed and the first desired.

daschuer committed:
Improve comment

daschuer committed:
Remove redundant constexpr on a local variable and improve comments

Literal constants are always constexpr and the defined local variables live on stack or is optimized away. Alternative: use static constexpr.

Swiftb0y reviewed (approved):
Since this fixes a bug, I won't hold this PR up over naming and codestyle discussions. LGTM.

daschuer commented:
Thank you.
